capsules/src/ltc294x.rs

```diff
@@ -0,0 +1,469 @@
+//! Driver for the LTC294X line of coloumb counters
+//!
+//! http://www.linear.com/product/LTC2941
+//! http://www.linear.com/product/LTC2942
+//! http://www.linear.com/product/LTC2943
+
+use core::cell::Cell;
+
+use kernel::{AppId, Callback, Driver};
+use kernel::common::take_cell::TakeCell;
+use kernel::hil::gpio;
+use kernel::hil::i2c;
+use kernel::returncode::ReturnCode;
+
+pub static mut BUFFER: [u8; 20] = [0; 20];
+
+#[allow(dead_code)]
+enum Registers {
+    Status = 0x00,
+    Control = 0x01,
+    AccumulatedChargeMSB = 0x02,
+    AccumulatedChargeLSB = 0x03,
+    ChargeThresholdHighMSB = 0x04,
+    ChargeThresholdHighLSB = 0x05,
+    ChargeThresholdLowMSB = 0x06,
+    ChargeThresholdLowLSB = 0x07,
+    VoltageMSB = 0x08,
+    VoltageLSB = 0x09,
+    CurrentMSB = 0x0E,
+    CurrentLSB = 0x0F,
+}
+
+#[derive(Clone,Copy,PartialEq)]
+enum State {
+    Idle,
+
+    /// Simple read states
+    ReadStatus,
+    ReadCharge,
+    ReadVoltage,
+    ReadCurrent,
+    ReadShutdown,
+
+    Done,
+}
+
+#[derive(Clone,Copy)]
+pub enum ChipModel {
+    LTC2941 = 1,
+    LTC2942 = 2,
+    LTC2943 = 3,
+}
+
+pub enum InterruptPinConf {
+    Disabled = 0x00,
+    ChargeCompleteMode = 0x01,
+    AlertMode = 0x02,
+}
+
+pub enum VBatAlert {
+    Off = 0x00,
+    Threshold2V8 = 0x01,
+    Threshold2V9 = 0x02,
+    Threshold3V0 = 0x03,
+}
+
+pub trait LTC294XClient {
+    fn interrupt(&self);
+    fn status(&self,
+              undervolt_lockout: bool,
+              vbat_alert: bool,
+              charge_alert_low: bool,
+              charge_alert_high: bool,
+              accumulated_charge_overflow: bool);
+    fn charge(&self, charge: u16);
+    fn voltage(&self, voltage: u16);
+    fn current(&self, current: u16);
+    fn done(&self);
+}
+
+pub struct LTC294X<'a> {
+    i2c: &'a i2c::I2CDevice,
+    interrupt_pin: Option<&'a gpio::Pin>,
+    model: Cell<ChipModel>,
+    state: Cell<State>,
+    buffer: TakeCell<'static, [u8]>,
+    client: Cell<Option<&'static LTC294XClient>>,
+}
+
+impl<'a> LTC294X<'a> {
+    pub fn new(i2c: &'a i2c::I2CDevice,
+               interrupt_pin: Option<&'a gpio::Pin>,
+               buffer: &'static mut [u8])
+               -> LTC294X<'a> {
+        LTC294X {
+            i2c: i2c,
+            interrupt_pin: interrupt_pin,
+            model: Cell::new(ChipModel::LTC2941),
```

Comment:
Sort of? Everything that you can do to the 41 won't fail on the 42/43, but some of the bit fields have different meanings. Seems fine to just default to the simple one.